workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/SpatialOSBehaviourManager.cs

```diff
@@ -1,60 +1,238 @@
+            = new Dictionary<uint, HashSet<MonoBehaviour>>();
+
+        private readonly Dictionary<uint, HashSet<MonoBehaviour>> componentIdToWriterSpatialOSBehaviours
+            = new Dictionary<uint, HashSet<MonoBehaviour>>();
+
+        private readonly Dictionary<MonoBehaviour, int> numUnSatisfiedReadersWriters
+            = new Dictionary<MonoBehaviour, int>();
+
+        private readonly Dictionary<MonoBehaviour, Dictionary<uint, IReaderInternal>> behaviourToCreatedReadersWriters
+            = new Dictionary<MonoBehaviour, Dictionary<uint, IReaderInternal>>();
+
+        private readonly HashSet<MonoBehaviour> spatialOSBehavioursToEnable = new HashSet<MonoBehaviour>();
+        private readonly HashSet<MonoBehaviour> spatialOSBehavioursToDisable = new HashSet<MonoBehaviour>();
+
+        private readonly Dictionary<uint, HashSet<IReaderInternal>> componentIdToCreatedReadersWriters =
+            new Dictionary<uint, HashSet<IReaderInternal>>();
 
-        public SpatialOSBehaviourManager(GameObject gameObject)
+        private readonly SpatialOSBehaviourLibrary behaviourLibrary;
+
+        private readonly ILogDispatcher logger;
+
+        private const string NotInterestedInComponent =
+            "SpatialOSBehaviourManager received update on component it is not interested in!";
+
+        private const string LoggerName = "GameObjectReference";
+
+        public SpatialOSBehaviourManager(GameObject gameObject, SpatialOSBehaviourLibrary library, ILogDispatcher logger)
         {
-            // Setup caches
+            this.logger = logger;
+
+            behaviourLibrary = library;
+            var spatialComponent = gameObject.GetComponent<SpatialOSComponent>();
+            entity = spatialComponent.Entity;
+            spatialId = spatialComponent.SpatialEntityId;
+            foreach (var behaviour in gameObject.GetComponents<MonoBehaviour>())
+            {
+                var readerIds = library.GetRequiredReaderComponentIds(behaviour.GetType());
+                foreach (var id in readerIds)
+                {
+                    if (!componentIdToReaderOrWriterSpatialOSBehaviours.ContainsKey(id))
+                    {
+                        componentIdToReaderOrWriterSpatialOSBehaviours.Add(id, new HashSet<MonoBehaviour>());
+                    }
+
+                    componentIdToReaderOrWriterSpatialOSBehaviours[id].Add(behaviour);
+                }
+
+                var writerIds = library.GetRequiredWriterComponentIds(behaviour.GetType());
+                foreach (var id in writerIds)
+                {
+                    if (!componentIdToWriterSpatialOSBehaviours.ContainsKey(id))
+                    {
+                        componentIdToWriterSpatialOSBehaviours.Add(id, new HashSet<MonoBehaviour>());
+                    }
+
+                    componentIdToWriterSpatialOSBehaviours[id].Add(behaviour);
```

Comment:
Do we need to add it to the `componentIdToReaderOrWriterSpatialOSBehaviours` dictionary too?

Reply:
Yes it should have been like that, but after looking at it more I'm actually very confused about why I even wanted to iterate over that set of behaviours at any point. The only place I'm doing that is when I add/remove components, but that is incorrect since writers should be marked as ready based on authority change, not adding/removing components... So I'll change this by making it just store readers in that map (also makes isReaderWanted checks cleaner).